Add --ignore flag to ignore files/directories

takos22 opened the pull request:
Add --ignore flag to ignore files/directories
Alias `-I`, can be used multiple times

mgedmin commented on findimports.py:
```diff
@@ -80,7 +80,7 @@
 
 
-__version__ = '1.5.3.dev0'
+__version__ = '1.5.3.dev1'
```
It is not necessary to increment the `.dev` suffix, but since it's a new feature, SemVer would suggest incrementing the minor version number:
```suggestion
__version__ = '1.6.0.dev0'
```

mgedmin commented on findimports.py:
```diff
@@ -418,15 +418,23 @@ def warn(self, about, message, *args):
                 dirs.sort()
                 files.sort()
+
+                for ignore in ignores:
```
If `ignores` is `None`, this will fail.

mgedmin replied:
Since the code itself doesn't allow `None` to be passed, I'd suggest changing the method signature to make `ignores` a positional parameter with no default value.

mgedmin replied:
Oh, but I see tests currently don't pass the extra argument.

In which case it may be simplest to 
```suggestion
                for ignore in ignores or []:
```

mgedmin commented on findimports.py:
```diff
@@ -418,15 +418,23 @@ def warn(self, about, message, *args):
+                    if ignore in dirs:
+                        dirs.remove(ignore)
+                    if ignore in files:
+                        files.remove(ignore)
```
Hm, no wildcard support?

mgedmin replied:
(We can add that later, if there's demand.)

mgedmin commented on findimports.py:
```diff
@@ -848,7 +859,7 @@ def main(argv=None):
     g.trackUnusedNames = (opts.action == 'printUnusedImports')
     for fn in args:
-        g.parsePathname(fn)
+        g.parsePathname(fn, ignores=opts.ignore or ["venv"])
```
If there's a default, it would be nice if `--help` mentioned it.

mgedmin reviewed (changes requested):
Thank you, this looks good!  For this to be merged, I'd like to see some tests for the new feature, and maybe a one-line note in the changelog.

takos22 commented:
Done the changes

mgedmin commented:
Excellent!  Now the only missing thing is unit tests for the new code:

```
Name             Stmts   Miss  Cover   Missing
----------------------------------------------
findimports.py     525      2    99%   434, 436
----------------------------------------------
TOTAL              525      2    99%
Coverage failure: total of 100 is less than fail-under=100
```

It would be easier to write the tests after some refactoring of the code.  I would extract the filtering logic into a separate method:

```
    def filterIgnoes(self, dirs, files, ignores):
                for ignore in ignores:
                    if ignore in dirs:
                        dirs.remove(ignore)
                    if ignore in files:
                        files.remove(ignore)
```

call it inside parsePathname(), and then write a test like

```
    def test_filterIgnores(self):
        dirs = ['venv', 'submodule']
        files = ['code.py', 'README.txt']
        mg = findImports.ModuleGraph()
        mg.filterIgnores(dirs, files, ignores=['venv', 'README.txt'])
        self.assertEqual(dirs, ['submodule'])
        self.assertEqual(files, ['code.py'])
```

in tests.py, next to the parsePathname test.

takos22 commented:
I was busy lately but I will try to make the changes tomorrow.

mgedmin commented:
There's no rush! ;)

takos22 commented:
Separated the logic in `ModuleGraph.filterIgnores`, added tests, updated the docstring in findimports.py, bumped the version to 0.2.1 and added a short description in `CHANGES.rst`. Could you review again?

mgedmin commented on findimports.py:
```diff
@@ -7,27 +7,37 @@
+                        Ignore a file or directory; this option can be used
+                        multiple times. Default: ['venv']
+
+FindImports requires Python 3.6 or later.
```
Oh wow, and I'd totally forgotten that there was a usage message in the docstring!  Thank you for noticing and updating it!

(This code is _old_, these days I'd rely on argparse-generated help messages, and maybe put a copy in the README for new user convenience, and then usually forget to update it when making changes.)

takos22 replied:
I'm making modules too so I know the feeling when I forget to update docs 😅

mgedmin reviewed (approved):
Looks good to me!

mgedmin commented:
I've pushed a new release to PyPI.
